contracts/sfc/SubsidiesRegistry.sol

```diff
@@ -0,0 +1,264 @@
+        SFC.burnNativeTokens{value: fee}();
+    }
+
+    /// @notice Insert funds into a specified sponsorship fund.
+    /// @dev Increases the fund's available balance, the sender's contribution record,
+    ///      and the total contributions counter.
+    /// @param fundId The unique identifier of the sponsorship fund.
+    function sponsor(bytes32 fundId) public payable {
+        require(fundId != bytes32(0), ZeroFundId());
+        require(msg.value > 0, NoFundsAttached());
+        require(owner() != address(0), NotInitialized()); // avoid paying into implementation contract
+
+        Fund storage fund = sponsorships[fundId];
+        fund.available += msg.value;
+        fund.contributors[msg.sender] += msg.value;
+        fund.totalContributions += msg.value;
+
+        emit Sponsored(fundId, msg.sender, msg.value);
+    }
+
+    /// @notice Withdraw funds from a sponsorship fund.
+    /// @dev A sponsor can withdraw up to their proportional share of the fund's available balance.
+    ///      Withdrawals are blocked in sponsored transactions to prevent abuse.
+    /// @param fundId The unique identifier of the sponsorship fund.
+    /// @param amount The requested withdrawal amount (in wei). If larger than the maximum allowed,
+    ///        it will be capped to the sponsor's withdrawable share.
+    function withdraw(bytes32 fundId, uint256 amount) external {
+        require(fundId != bytes32(0), ZeroFundId());
+        Fund storage fund = sponsorships[fundId];
+        uint256 maxAmount = _availableToWithdraw(fund, msg.sender);
+        if (amount > maxAmount) amount = maxAmount;
+        require(amount != 0, NothingToWithdraw());
+        require(tx.gasprice != 0, NotAllowedInSponsoredTx());
+
+        uint256 toSubstract = (amount * fund.totalContributions) / fund.available;
+        fund.available -= amount;
+        fund.contributors[msg.sender] -= toSubstract;
+        fund.totalContributions -= toSubstract;
+
+        (bool success, ) = msg.sender.call{value: amount}("");
+        require(success, TransferFailed());
+
+        emit Withdrawn(fundId, msg.sender, amount);
+    }
+
+    /// @notice Get amount of funds available for a sponsor to withdraw from a fund
+    /// @param fundId The fund to withdrawn
+    /// @param _sponsor The withdrawing sponsor
+    function availableToWithdraw(bytes32 fundId, address _sponsor) external view returns (uint256) {
+        Fund storage fund = sponsorships[fundId];
+        return _availableToWithdraw(fund, _sponsor);
+    }
+
+    /// @notice Get amount of funds available for a sponsor to withdraw from a fund
+    /// @param fund The fund to withdrawn
+    /// @param _sponsor The withdrawing sponsor
+    function _availableToWithdraw(Fund storage fund, address _sponsor) private view returns (uint256) {
+        if (fund.totalContributions == 0) {
+            return 0;
+        }
+        return (fund.available * fund.contributors[_sponsor]) / fund.totalContributions;
```

Comment:
Do we have to worry about overflows in the multiplication (also in the "withdraw" function above)?